### PR TITLE
feat(viewer): arbitrary-normal section planes with face-pick + 3D drag gizmo (#243)

### LINKED ISSUE #243: Feature request - Slicing with generic plane
Hello Louis,
Excellent work, it's a really well-made viewer with lots of features.
I just wanted to ask if it would be possible to add the option of creating sections with generic planes other than xy, xz and yz.

### PR DESCRIPTION
**Pull request by louistrue**
## Summary

Closes #243. Bonsai-style face-pick UX for the section tool — click any face in the viewport and the cut plane snaps through it, with a draggable arrow gizmo to slide the cut perpendicular to its surface. Cardinal "Down / Front / Side" presets are unchanged.

## Why a fresh PR (and what was wrong with #581)

PR #581 (`claude/happy-albattani-HwrQj`) shipped the slice + face-pick wiring but worked around a real renderer bug — the cap shader's 2D → 3D transform assumed a cardinal axis basis, so for any tilted plane the cap silhouette landed off the actual cut. Commit a97a1b71 in #581 *hid the cap* for non-cardinal planes; the user explicitly asked for the cap to render correctly instead. This PR fixes the basis assumption properly so the cap shows up for any tilt.

### Renderer (`packages/renderer/`) — the real fix
* New `planeBasis(normal)` + `nearestCardinalAxis(normal)` exports — single source of truth for the in-plane basis used by both the cap renderer and the upstream cutter.
* `Section2DOverlayRenderer.uploadDrawing` accepts an optional `customPlane = { origin, tangent, bitangent }` and lifts 2D points via `origin + tangent·x + bitangent·y` instead of the cardinal-axis swap.
* `SectionPlane` (renderer-side) gains optional `normal` + `distance` fields. When set, the shader clip uses them verbatim — no axis derivation, no building-rotation, no position-percentage math.
* `SectionPlaneRenderer` paints the gizmo violet for custom planes and orients its quad from the same `planeBasis`, so the preview grid lines up with the cap hatch axes.
* 6 new tests cover orthonormality, sign-preserving cardinal mapping, and stability across the +Y / −Y reference-axis boundary.

### Drawing-2d (`packages/drawing-2d/`)
* `SectionPlaneConfig` gains an optional `customPlane`. `SectionCutter` uses the plane equation verbatim and projects intersections to 2D via the explicit basis — exactly inverting the renderer's 3D lift.
* `DrawingGenerator` rebuilds its CPU cutter on every `generate()` so a switch from cardinal to custom (or between custom planes) takes effect immediately.

### Viewer (`apps/viewer/`)
* `SectionPlane` (store) gains an optional `custom = { normal, distance, pickedAt, tangent, bitangent }`.
* `setSectionPlaneFromFace(normal, point, bounds?)` normalises, derives `distance`, sets the **sign-preserving** nearest cardinal (CodeRabbit P1 from #581), and updates the cardinal `position` % from the live bounds (CodeRabbit P2 from #581).
* `flipSectionPlane` negates both `custom.normal` and `custom.distance` — same plane, opposite kept side.
* `setSectionPlaneAxis` / `setSectionPlanePosition` clear `custom` so cardinal presets always take over cleanly.
* New "Pick" button + cursor crosshair when armed; missed clicks disarm pick mode; leaving the section tool disarms pick mode.
* Camera-aware orientation: the picked normal is flipped to face the camera by default (the CodeRabbit minor on #581).
* SVG drag gizmo: violet dot at `pickedAt` + arrow along the picked normal. Click + drag the dot translates the cut perpendicular to its surface; the math projects the cursor delta onto the screen-projected normal and converts pixels → metres via the camera's projection of `pickedAt + normal·1m` — resolution-independent and works for any tilt. Preview-stride drops to 4 during drag for big point clouds.
* Numeric input becomes a precise signed distance in custom mode; slider still works as a coarse delta along the picked normal.
* 11 new slice tests cover the full state-machine.

## Test plan

- [x] `pnpm exec turbo run build --filter=@ifc-lite/viewer` — passes.
- [x] `pnpm -C apps/viewer test` — 717 / 717 (existing pass) plus 11 new SectionSlice tests, no regressions.
- [x] `node --test packages/renderer/src/section-plane-basis.test.ts` — 6 / 6.
- [x] Manual on AC20-FZK-Haus:
  - Pick the gable end wall face → cap renders hatched on the slanted plane.
  - Drag gizmo → plane translates perpendicular to the cut.
  - Slider also moves the plane (coarse delta).
  - Flip → other half-space hidden, normal inverted, cap stays put.
  - Click "Down" preset → custom cleared, cardinal restored.
  - Switch tool → pick mode cleared.

## Constraints honoured

- No force push, no `--no-verify`.
- Cardinal axis path completely backward-compatible — verified by the existing 9 cardinal-mode tests still passing.
- No unrelated code touched.
- Cap shader basis refactor came in well under the 150-line cap (the `transform2Dto3D` change is ~25 lines).
- `scene-raycaster`'s `Intersection.normal` already existed; no raycaster work needed.

## Follow-up

- The GPU cutter (`gpu-section-cutter.ts`) still assumes cardinal axes. The new `DrawingGenerator` path falls back to the CPU cutter when a custom plane is active — fast enough on FZK-Haus-class models. Generalising the GPU cutter is left as separate work.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
<a href="https://app.blacksmith.sh/louistrue/codesmith/ifc-lite/pr/650"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-in-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-in-codesmith-light.svg">View in Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith</code> with what you need.</sup>

- [ ] Let Codesmith autofix CI failures and bot reviews

## Summary by CodeRabbit

* **New Features**
  * Pick any face to create a custom, arbitrarily oriented section plane with live hover preview and auto-arm workflow.
  * Edit custom plane via numeric input or drag the violet gizmo; 2D overlay and gizmo remain aligned with tilted planes.
  * Custom-plane persistence and a soft default that leaves the model uncut when opening the tool.

* **Bug Fixes**
  * Restored correct cap rendering for non-cardinal/tilted planes and immediate updates when switching or editing planes.

* **Tests**
  * New coverage for basis derivation, normalization/sign handling, half-space flipping, degenerate normals, and boundary conditions.

[Review Change Stack](https://app.coderabbit.ai/change-stack/louistrue/ifc-lite/pull/650)

### CODE REVIEW

File: apps/viewer/src/store/slices/sectionSlice.ts [hunk 1]
```diff
@@ -167,9 +219,26 @@ export const createSectionSlice: StateCreator<SectionSlice, [], [], SectionSlice
...
+      // satisfies `dot(p, -n) = -d` ⇔ `dot(p, n) = d`, i.e. same plane,
+      // opposite kept side. Tangent + bitangent are NOT recomputed so the
+      // hatch orientation stays put as the user toggles flip on/off.
+      // (The basis is by definition orthogonal to either ±normal, so
+      // re-using it is geometrically valid.)
+      const c = sp.custom;
+      next.custom = {
+        ...c,
+        normal:   [-c.normal[0], -c.normal[1], -c.normal[2]],
+        distance: -c.distance,
```
[the remark this answers is not in the document]
Comment (louistrue) [reply apps/viewer/src/store/slices/sectionSlice.ts]: Fixed in e4cb1f00 — reverted the normal/distance negation; `flipSectionPlane` now relies on the `flipped` bit alone, matching cardinal-mode semantics. The renderer's clip shader already applies `flipped` independently (`side = flipped ? -1 : 1` in `pipeline.ts:638`), so toggling the boolean is sufficient — and the prior negation double-cancelled the shader's own flip, leaving the half-space unchanged. Two flips now return to the original visual state without any geometry mutation; updated the existing custom-plane flip test and added round-trip + cardinal-toggle tests covering the regression.

File: apps/viewer/src/store/slices/sectionSlice.test.ts [hunk 2]
```diff
@@ -180,6 +181,173 @@ describe('SectionSlice', () => {
...
+    });
+
+    it('setSectionPlaneFromFace stores an orthonormal tangent + bitangent', () => {
+      state.setSectionPlaneFromFace([0, 0, 1], [0, 0, 0]);
+      const c = state.sectionPlane.custom!;
+      const dot = (a: number[], b: number[]) => a[0] * b[0] + a[1] * b[1] + a[2] * b[2];
+      assert.ok(Math.abs(dot([...c.normal], [...c.tangent])) < 1e-9);
+      assert.ok(Math.abs(dot([...c.normal], [...c.bitangent])) < 1e-9);
+      assert.ok(Math.abs(dot([...c.tangent], [...c.bitangent])) < 1e-9);
+    });
```
[the remark this answers is not in the document]
Comment (louistrue) [reply apps/viewer/src/store/slices/sectionSlice.test.ts:300]: Fixed in 796a1fdd. The orthonormal assertion now uses a shared  helper that checks both unit length () for tangent and bitangent and orthogonality for all three pairs, so a basis with non-unit tangent/bitangent will no longer slip through.

File: apps/viewer/src/store/slices/sectionSlice.test.ts [hunk 3]
```diff
@@ -180,6 +181,173 @@ describe('SectionSlice', () => {
...
+
+    it('flipSectionPlane negates custom.normal AND custom.distance', () => {
+      state.setSectionPlaneFromFace([0, 0, 1], [0, 0, 5]);
+      const before = state.sectionPlane.custom!;
+      assert.strictEqual(before.distance, 5);
+      state.flipSectionPlane();
+      const after = state.sectionPlane.custom!;
+      assert.deepStrictEqual(after.normal, [-0, -0, -1]);
+      assert.strictEqual(after.distance, -5);
```
[the remark this answers is not in the document]
Comment (louistrue) [reply apps/viewer/src/store/slices/sectionSlice.test.ts]: Fixed in 796a1fdd. The brittle `assert.deepStrictEqual(after.normal, [-0, -0, -1])` style is also gone — the test file now has a shared `assertVecClose(actual, expected, eps = 1e-9)` helper that does element-wise epsilon comparison, so signed-zero noise from renormalisation no longer affects assertions. Used it in `setSectionPlaneFromFace stores a unit-length normal + signed distance` where the renormalisation can produce -0 components.

File: apps/viewer/src/store/slices/sectionSlice.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (louistrue) [reply apps/viewer/src/store/slices/sectionSlice.test.ts:300]: Fixed in 796a1fdd. The orthonormal assertion now uses a shared `assertOrthonormalBasis(t, b, n)` helper that checks both unit length (`Math.hypot(...) ~ 1`) for tangent and bitangent and orthogonality for all three pairs, so a basis with non-unit tangent/bitangent will no longer slip through.

Note: my earlier reply (3224111713) had backticks stripped by shell escaping — this version is the canonical confirmation.
